File: src/main/java/tech/wintrich/networkapi/application/service/HttpAnalysisService.java
package tech.wintrich.networkapi.application.service;

import lombok.extern.slf4j.Slf4j;
import org.springframework.beans.factory.annotation.Value;
import org.springframework.stereotype.Service;
import tech.wintrich.networkapi.presentation.dto.HttpAnalysisResponse;

import java.io.IOException;
import java.net.URI;
import java.net.http.HttpClient;
import java.net.http.HttpRequest;
import java.net.http.HttpResponse;
import java.time.Duration;
import java.util.*;
import java.util.stream.Collectors;

/**
 * Service for HTTP analysis and performance testing
 */
@Slf4j
@Service
public class HttpAnalysisService {

    @Value("${api.network.timeout.http:10000}")
    private int httpTimeout;

    private final HttpClient httpClient;

    public HttpAnalysisService() {
        this.httpClient = HttpClient.newBuilder()
                .followRedirects(HttpClient.Redirect.NORMAL)
                .build();
    }

    /**
     * Performs comprehensive HTTP analysis of a URL
     *
     * @param url the target URL
     * @return HttpAnalysisResponse with performance and header data
     */
    public HttpAnalysisResponse analyze(String url) {
        log.debug("Performing HTTP analysis for URL: {}", url);

        HttpAnalysisResponse.HttpAnalysisResponseBuilder responseBuilder = 
                HttpAnalysisResponse.builder().url(url);

        try {
            // Build request with timeout
            HttpRequest request = HttpRequest.newBuilder()
                    .uri(URI.create(url))
                    .timeout(Duration.ofMillis(httpTimeout))
                    .GET()
                    .build();

            // Measure response time
            long startTime = System.nanoTime();
            HttpResponse<String> response = httpClient.send(request, 
                    HttpResponse.BodyHandlers.ofString());
            long responseTime = (System.nanoTime() - startTime) / 1_000_000;

            // Extract response data
            responseBuilder
                    .status(response.statusCode())
                    .statusText(getStatusText(response.statusCode()))
                    .responseTimeMs(responseTime)
                    .finalUrl(response.uri().toString());

            // Process headers
            Map<String, String> headers = response.headers().map().entrySet().stream()
                    .collect(Collectors.toMap(
                            Map.Entry::getKey,
                            e -> String.join(", ", e.getValue())
                    ));

            responseBuilder.headers(headers);

            // Extract specific headers
            response.headers().firstValue("content-type")
                    .ifPresent(responseBuilder::contentType);

            response.headers().firstValue("server")
                    .ifPresent(responseBuilder::server);

            response.headers().firstValue("content-length")
                    .ifPresent(length -> responseBuilder.contentLength(Long.parseLong(length)));

            // Check for redirects
            if (!response.uri().toString().equals(url)) {
                responseBuilder.redirectChain(new String[]{url, response.uri().toString()});
            }

            // If no content-length header, use body length
            if (responseBuilder.build().getContentLength() == null) {
                responseBuilder.contentLength((long) response.body().length());
            }

        } catch (IOException e) {
            log.error("HTTP request failed for URL: {}", url, e);
            responseBuilder.error("HTTP request failed: " + e.getMessage());
        } catch (InterruptedException e) {
            log.error("HTTP request interrupted for URL: {}", url, e);
            Thread.currentThread().interrupt();
            responseBuilder.error("Request interrupted");
        } catch (Exception e) {
            log.error("Unexpected error during HTTP analysis: {}", url, e);
            responseBuilder.error("Analysis failed: " + e.getMessage());
        }

        return responseBuilder.build();
    }

    /**
     * Converts HTTP status code to text
     */
    private String getStatusText(int statusCode) {
        return switch (statusCode) {
            case 200 -> "OK";
            case 201 -> "Created";
            case 204 -> "No Content";
            case 301 -> "Moved Permanently";
            case 302 -> "Found";
            case 304 -> "Not Modified";
            case 400 -> "Bad Request";
            case 401 -> "Unauthorized";
            case 403 -> "Forbidden";
            case 404 -> "Not Found";
            case 500 -> "Internal Server Error";
            case 502 -> "Bad Gateway";
            case 503 -> "Service Unavailable";
            case 504 -> "Gateway Timeout";
            default -> "HTTP " + statusCode;
        };
    }
}

File: web-api/Services/HttpAnalysisService.cs
using Wintrich.NetworkApi.DTOs;

namespace Wintrich.NetworkApi.Services;

/// <summary>
/// HTTP analysis service using IHttpClientFactory.
/// Equivalent of the Java HttpAnalysisService (measures response time, captures headers, redirects).
/// </summary>
public sealed class HttpAnalysisService(
    IHttpClientFactory httpClientFactory,
    ILogger<HttpAnalysisService> logger)
{
    public async Task<HttpAnalysisResponse> AnalyzeAsync(string url)
    {
        logger.LogDebug("Performing HTTP analysis for URL: {Url}", url);

        int? status = null;
        string? statusText = null;
        long? responseTimeMs = null;
        string? contentType = null;
        string? server = null;
        Dictionary<string, string>? headers = null;
        string[]? redirectChain = null;
        string? finalUrl = null;
        long? contentLength = null;
        string? error = null;

        try
        {
            var client = httpClientFactory.CreateClient("analysis");

            var sw = System.Diagnostics.Stopwatch.StartNew();
            using var response = await client.GetAsync(url, HttpCompletionOption.ResponseContentRead);
            sw.Stop();

            responseTimeMs = sw.ElapsedMilliseconds;
            status = (int)response.StatusCode;
            statusText = GetStatusText(status.Value);
            finalUrl = response.RequestMessage?.RequestUri?.ToString() ?? url;

            // Collect all response headers (flatten multi-value headers with ", ")
            headers = response.Headers
                .Concat(response.Content.Headers)
                .ToDictionary(
                    h => h.Key.ToLowerInvariant(),
                    h => string.Join(", ", h.Value),
                    StringComparer.OrdinalIgnoreCase);

            contentType = response.Content.Headers.ContentType?.ToString();
            server = response.Headers.Server?.ToString();

            // Content length: header first, then actual body length as fallback
            contentLength = response.Content.Headers.ContentLength;
            if (contentLength == null)
            {
                var body = await response.Content.ReadAsByteArrayAsync();
                contentLength = body.LongLength;
            }

            // Redirect chain: if the final URL differs from the original
            if (!string.Equals(finalUrl, url, StringComparison.OrdinalIgnoreCase))
                redirectChain = new[] { url, finalUrl };
        }
        catch (HttpRequestException ex)
        {
            logger.LogError(ex, "HTTP request failed for URL: {Url}", url);
            error = $"HTTP request failed: {ex.Message}";
        }
        catch (TaskCanceledException ex)
        {
            logger.LogError(ex, "HTTP request timed out for URL: {Url}", url);
            error = "Request timed out";
        }
        catch (Exception ex)
        {
            logger.LogError(ex, "Unexpected error during HTTP analysis for URL: {Url}", url);
            error = $"Analysis failed: {ex.Message}";
        }

        return new HttpAnalysisResponse(
            url, status, statusText, responseTimeMs,
            contentType, server, headers, redirectChain, finalUrl, contentLength, error);
    }

    private static string GetStatusText(int statusCode) => statusCode switch
    {
        200 => "OK",
        201 => "Created",
        204 => "No Content",
        301 => "Moved Permanently",
        302 => "Found",
        304 => "Not Modified",
        400 => "Bad Request",
        401 => "Unauthorized",
        403 => "Forbidden",
        404 => "Not Found",
        500 => "Internal Server Error",
        502 => "Bad Gateway",
        503 => "Service Unavailable",
        504 => "Gateway Timeout",
        _ => $"HTTP {statusCode}"
    };
}
